Feature: Add DAO profiles

Evalir opened the pull request:
Adds DAO profiles to Apiary, letting users authenticate through web3 & their DAO's ACL to edit profiles stored in the database.

This has been tested both on Rinkeby & Mainnet in a very thorough manner. For reviewers, the `AccountModule` and `assets` parts of this PR were extracted from the court dashboard as-is and integrated with Apiary.

I'll do some things today to let it fully ready to merge:
- [x] Refactor to comply with stablished coding standards (solve all warnings from standard.js)
- [x] More documentation across all changes.
- [ ] Update README.md with changes.

Some screenshots:

<img width="1680" alt="Screen Shot 2020-04-29 at 4 11 25 AM" src="https://user-images.githubusercontent.com/26014927/80574279-8cb30700-89cf-11ea-8117-11d449bdafcb.png">
<img width="1680" alt="Screen Shot 2020-04-29 at 4 11 17 AM" src="https://user-images.githubusercontent.com/26014927/80574285-8fadf780-89cf-11ea-9195-852764e7e7d9.png">
<img width="1680" alt="Screen Shot 2020-04-29 at 4 07 44 AM" src="https://user-images.githubusercontent.com/26014927/80574288-90468e00-89cf-11ea-9c10-90daf5fc5ead.png">

onbjerg commented on api/src/orgs/data.js:
```diff
@@ -39,6 +46,55 @@ export function getOrganisations (
+    return
+  }
+
+  const { profile = {} } = await db.collection('orgs').findOne({ address })
+  const currentEditors = profile.editors || []
+  const newEditors = [...currentEditors, signerAddress]
+  // Remove duplicates
+  const filteredEditors = new Set(newEditors)
+  const newFilteredEditors = [...filteredEditors]
```
I think this code would achieve the same result (we might want to rename `newFilteredEditors` to something more succinct in that case)

```suggestion
  const {
    profile = {}
  } = await db.collection('orgs').findOne({ address })
  const newFilteredEditors = new Set(profile.editors || [])
  newFilteredEditors.set(signerAddress)
```

Evalir replied:
Way nicer! Done in https://github.com/1Hive/apiary/pull/181/commits/fc57eb6088b2c7cafe0bdce21a4b14c03f6523eb.

onbjerg commented on api/src/orgs/data.js:
```diff
@@ -39,6 +46,55 @@ export function getOrganisations (
+  return db.collection('orgs').findOne({ address })
+}
+
+export async function getSingleOrganisation (
```
I think just `getOrganisation` is fine. Also there is no need for this function to be `async` since it returns a promise and does not use `await`

Evalir replied:
Done in https://github.com/1Hive/apiary/pull/181/commits/47385dfa29ed128f82c09f3045de9cbefc67de6f.

onbjerg commented on api/src/orgs/data.js:
```diff
@@ -39,6 +46,55 @@ export function getOrganisations (
+
+export async function getSingleOrganisation (
+  db,
+  args
+) {
+  const { address } = args
```
```suggestion
  { address }
) {
```

onbjerg commented on api/src/orgs/resolver.js:
```diff
@@ -1,4 +1,11 @@
-import { getOrganisations } from './data'
+import { getOrganisations, getSingleOrganisation, updateProfile } from './data'
+
+const EMPTY_PROFILE = {
```
It seems like defaults are already set in the `Profile` resolver a bit further down, so there is no need for `EMPTY_PROFILE`

Evalir replied:
Done in https://github.com/1Hive/apiary/pull/181/commits/6280fdcc5b5f13aae91b538428c8ec1d83e4ccc4.

onbjerg commented on api/src/orgs/resolver.js:
```diff
@@ -7,7 +14,41 @@ export default {
+  },
+
+  Profile: {
+    name (profile) {
+      return profile.name || ''
+    },
+    icon (profile) {
+      return profile.icon || ''
+    },
+    links (profile) {
+      return profile.links || []
+    },
+    editors (profile) {
+      return profile.editors || []
+    },
+    description (profile) {
+      return profile.description || ''
     }
```
In terms of defaults in this section, there is no reason to return empty strings for most of these fields since they are marked as nullable in the schema (i.e. no `!`), and as such I think they should return `null` (except for the arrays which are marked as non-nullable).

Evalir replied:
Done in https://github.com/1Hive/apiary/pull/181/commits/fe3ae6f8174afbe2d00ddbb555a6cb85eacf3228.

onbjerg commented on api/src/orgs/schema.js:
```diff
@@ -1,4 +1,18 @@
+    name: String
+    # URL to the organisation's icon.
+    icon: String
+    # Array of URL to the different links for the organisation (ex: Discord, Discourse)
```
```suggestion
    # Array of URL to the different links for the organisation (e.g. Discord, Discourse)
```

onbjerg commented on api/src/orgs/schema.js:
```diff
@@ -1,4 +1,18 @@
+    icon: String
+    # Array of URL to the different links for the organisation (ex: Discord, Discourse)
+    links: [String]!
+    # Array of addresses from the people that have edited the profile at least once
```
```suggestion
    # Array of addresses that have edited the profile at least once
```

I would omit the people part since it might not be a person ( 🤖 ) and it is a bit verbose I think

onbjerg commented on api/src/orgs/schema.js:
```diff
@@ -62,5 +78,23 @@ export default `
+  }
+
+  type Mutation {
+    # Update organisation profiles.
```
```suggestion
    # Update organisation profile.
```

onbjerg commented on api/src/orgs/schema.js:
```diff
@@ -62,5 +78,23 @@ export default `
+  type Mutation {
+    # Update organisation profiles.
+    updateProfile(
+      # The organization id.
```
```suggestion
      # The organization address.
```

onbjerg commented on api/src/utils.js:
```diff
@@ -11,3 +13,9 @@ export function camelToSnakeCaseKeys (obj = {}) {
   }, {})
 }
+
+export function composeSignedMessage (daoAddress, { name, description, links, icon }) {
```
Add a docstring or possibly rename to make it more clear that this is not a generic function to compose any signed message? Or possibly make it a bit more generic? Either one is fine 😊

Evalir replied:
Done in https://github.com/1Hive/apiary/pull/181/commits/9c7117bb63f9dc5bd0ac5964dc70176a621da71b.

onbjerg commented on website/package.json:
```diff
@@ -4,6 +4,8 @@
   "main": "index.html",
   "license": "GPLv3",
   "scripts": {
+    "start:mainnet": "CHAIN_ID=1 npm start",
+    "start:rinkeby": "CHAIN_ID=4 npm start",
```
We have some plans for multi-network support using the same front-end, so I think this should either be managed by the client themselves on the front-end somehow (possibly discovery) or by the API, if possible. (#14)

Evalir replied:
Yes; I'm not very sure it's possible right now with `use-wallet` for doing this; might get a bit contrived for now, as we need to set a provider with a chain ID. Can we find a workaround for this later?

onbjerg replied:
Yeah that's fine.

onbjerg commented on website/package.json:
```diff
@@ -14,6 +16,9 @@
   "dependencies": {
     "@aragon/ui": "^1.0.0",
+    "@aragon/wrapper": "^5.0.0-rc.26",
+    "clipboard-polyfill": "^2.8.6",
+    "crypto-js": "^4.0.0",
```
What are these used for?

Evalir replied:
`clipboard-polyfill` is used by the `AccountModule` for copying the address to the clipboard; the `crypto-js` library is for being able to take the `isAddress()` [implementation](https://github.com/Evalir/apiary/blob/master/website/src/utils/web3-utils.js#L64) from `web3`:

onbjerg commented on website/package.json:
```diff
@@ -14,6 +16,9 @@
   },
   "dependencies": {
     "@aragon/ui": "^1.0.0",
+    "@aragon/wrapper": "^5.0.0-rc.26",
```
Is it possible to use a preview version of the toolkit instead for transaction pathing? Even if it is not super stable in terms of API I think its fine for us if we just pin to a very specific version, given that the transaction pathing is sort of mature for this use case.

Evalir replied:
Right now, there's no preview version of the toolkit available for doing this (I tried), but we will migrate this to toolkit as soon as it's available.

onbjerg commented on website/src/components/Account/AccountModule.js:
```diff
@@ -0,0 +1,249 @@
+const AnimDiv = animated.div
+
+const SCREENS = [
+  {
```
Can you condense this array a bit more? I know its kind of a nit, but usually array of objects in the code base are written as

```js
const something = [{
  foo: 'bar'
}, {
  bar: 'baz'
}]
```

Evalir replied:
Done in https://github.com/1Hive/apiary/pull/181/commits/9fc22511a86e8d560ecf888b640de76873a24829.

onbjerg commented on website/src/components/LocalIdentityBadge/LocalIdentityBadge.js:
```diff
@@ -0,0 +1,15 @@
+}
+
+LocalIdentityBadge.propTypes = {
+  ...IdentityBadge.propTypes
```
If there are no additional props I think just doing `LocalIdentityBadge.propTypes = IdentityBadge.propTypes` is fine

Evalir replied:
Done in https://github.com/1Hive/apiary/pull/181/commits/e43aa46bf510777197f64544d278b58117c8d3f2.

onbjerg commented on website/src/components/SmartLink/SmartLink.js:
```diff
@@ -0,0 +1,24 @@
+import React, { useMemo } from 'react'
+
+const KNOWN_PLATFORMS = new Map([
+  ['discord', 'Discord'],
+  ['discourse', 'Discourse'],
+  ['github', 'Github'],
+  ['metacartel', 'Metacartel'],
+  ['twitter', 'Twitter'],
+  ['instagram', 'Instagram']
+])
+
+export default function SmartLink ({ url, ...props }) {
+  const title = useMemo(() => {
+    for (const [platformKey, platform] of KNOWN_PLATFORMS.entries()) {
+      if (url.includes(platformKey)) {
+        return platform
+      }
+    }
+
+    return url
+  }, [url])
+
+  return <a href={url} {...props}>{title}</a>
+}
```
Very clever and a nice touch! ✨

onbjerg commented on website/src/components/WindowedPagination/WindowedPagination.js:
```diff
@@ -19,7 +19,8 @@ export const WindowedPagination = React.memo(({
     width: 100%;
     margin: 1em 0;
-  '>
+  '
+  >
```
I think this is a mistake, no?

Evalir replied:
Yeah, seems to be, my bad. Done in https://github.com/1Hive/apiary/pull/181/commits/5b75dfab2dea2611ec54dee8bcee7008a4dd64aa.

onbjerg commented on website/src/ethereum-providers/index.js:
```diff
@@ -0,0 +1,114 @@
+import portis from './icons/Portis.svg'
+
+// See the corresponding prop type, EthereumProviderType, in prop-types.js.
+const PROVIDERS = new Map(
```
Same comment as for the other array of objects, if you could condense this I would appreciate it

Evalir replied:
Done in https://github.com/1Hive/apiary/pull/181/commits/9fc22511a86e8d560ecf888b640de76873a24829.

onbjerg commented on website/src/hooks/useCopyToClipboard.js:
```diff
@@ -0,0 +1,16 @@
+import { useToast } from '@aragon/ui'
+import { writeText as copy } from 'clipboard-polyfill'
+
+export function useCopyToClipboard () {
```
Is this not already a part of aUI? We already have some similar functionality by using the `IdentityBadge` for organisations in the overview.

Evalir replied:
Something like this is implemented but not exposed, and the way the AccountModule handles this is a bit different; eventually we will migrate this to aragonUI when we see another use case for this.

onbjerg commented on website/src/utils/ownership-statuses.js:
```diff
@@ -0,0 +1,12 @@
+export const OWNERSHIP_STATUSES = new Map([
+  ['CLAIM_PROFILE', 'Claim profile'],
+  ['EDIT_PROFILE', 'Edit profile'],
+  ['NOT_CONNECTED_PROFILE', 'Claim Profile'],
```
```suggestion
  ['NOT_CONNECTED_PROFILE', 'Claim profile'],
```

Adjusting text case to make sure they are the same style in all of the cases

onbjerg commented on website/src/utils/utils.js:
```diff
@@ -0,0 +1,46 @@
+export function constructPathDescription (transactionPath, proxies) {
+  const preTransactionPath = transactionPath.slice(0, transactionPath.length - 1)
+  const transactionDescription = preTransactionPath.map(({ description, to }, index) => {
+    //
```
```suggestion

```

Evalir replied:
Done in https://github.com/1Hive/apiary/pull/181/commits/1adad996efefbaa949de977d8508b6bbc2f7684f.

onbjerg commented on website/src/utils/utils.js:
```diff
@@ -0,0 +1,46 @@
+  })
+
+  const formattedDescription = <>{transactionDescription}</>
+  return [
+    {
+      title: 'Transaction Path',
+      description: formattedDescription
+    }
+  ]
```
Condense array of objects

Evalir replied:
Done in https://github.com/1Hive/apiary/pull/181/commits/9fc22511a86e8d560ecf888b640de76873a24829.

Evalir commented on api/src/orgs/data.js:
```diff
@@ -1,11 +1,18 @@
+import { validateSignerAddress } from '../web3-utils'
+
+const MANAGE_PROFILE_ROLE = '0x675b358b95ae7561136697fcc3302da54a334ac7c199d53621288290fb863f5c'
+const EMPTY_SCRIPT = '0x00'
```
Done in `https://github.com/1Hive/apiary/pull/181/commits/ee02e46f5b24a0079f96eb6c9f5c424b5c8f4c72`.

Evalir commented on api/src/orgs/data.js:
```diff
@@ -39,6 +46,55 @@ export function getOrganisations (
+
+  const originalMessage = composeSignedMessage(address, updateParams)
+  const isAddressValid = validateSignerAddress(originalMessage, signedMessage, signerAddress)
+  Web3EthContract.setProvider('wss://mainnet.eth.aragon.network/ws')
```
Done in https://github.com/1Hive/apiary/pull/181/commits/8af1486c770a06d1dad5ec5df5367f18f5729218. I also set the variable in the `docker-compose` file, as I think it's only set on the `heroku.yml` file.

Evalir commented on api/src/orgs/data.js:
```diff
@@ -39,6 +46,55 @@ export function getOrganisations (
+      MANAGE_PROFILE_ROLE,
+      EMPTY_SCRIPT
+    ).call()
+    if (!isAddressValid || !hasPermission) {
```
Indeed; done in https://github.com/1Hive/apiary/pull/181/commits/d089b700bab43ded535009358b3453dcc1fdd6bc.

onbjerg commented:
I forgot to mention, can we move the "Profile" heading in to the bar with the back button on it?

Evalir commented:
Thank you so much for the high quality review! Addressed the suggestions, and also took the time to:

1. Make some UX nicer on the `EditSidePanel` (character count).
2. Add `PropTypes` and alphabetize components.
3. Avoid overflow on the DataView if a user sets long links.

Evalir commented:
Note: decided to make the profiles non-nullable fields on the schema. While originally this was to reflect's the DB null state on the profiles, it does not makes sense for the API to return null; it should rather return an empty profile so the frontend can properly handle this case.

onbjerg commented:
Seems ready to be merged when the lint check passes. Do you have anything else you want to add beforehand?

Evalir commented:
Nothing else! Let me fix the lint check and let's get this merged. 💪
